### PR TITLE
SINGA-490 Optimize performance of stochastic gradient descent (SGD)

### PR DESCRIPTION
**Pull request by chrishkchris**
I have fused the small operations of momentum SGD so as to increase GPU computation efficiency and decrease latency. Moreover, I have added the Sync() function for better time profiling in resnet.py (wait for the previous cuda operations to be finished before start calculating the time).

1. This is the new result after improving the momentum SGD:

```
ubuntu@ip-172-31-39-137:~/incubator-singa/examples/autograd$ python3 mnist_cnn.py
Starting Epoch 0:
Training loss = 583.052124, training accuracy = 0.793690
Evaluation accuracy = 0.943409, Elapsed Time = 4.191409s
Starting Epoch 1:
Training loss = 229.894424, training accuracy = 0.923609
Evaluation accuracy = 0.961438, Elapsed Time = 4.170332s
Starting Epoch 2:
Training loss = 168.670303, training accuracy = 0.943937
Evaluation accuracy = 0.964744, Elapsed Time = 4.186504s
Starting Epoch 3:
Training loss = 133.865494, training accuracy = 0.955259
Evaluation accuracy = 0.978566, Elapsed Time = 4.188593s
Starting Epoch 4:
Training loss = 116.104378, training accuracy = 0.961730
Evaluation accuracy = 0.971554, Elapsed Time = 4.195830s
Starting Epoch 5:
Training loss = 101.295425, training accuracy = 0.966299
Evaluation accuracy = 0.974059, Elapsed Time = 4.191312s
Starting Epoch 6:
Training loss = 94.570869, training accuracy = 0.969684
Evaluation accuracy = 0.977464, Elapsed Time = 4.181115s
Starting Epoch 7:
Training loss = 85.930618, training accuracy = 0.970968
Evaluation accuracy = 0.984675, Elapsed Time = 4.182598s
Starting Epoch 8:
Training loss = 83.169617, training accuracy = 0.971768
Evaluation accuracy = 0.985076, Elapsed Time = 4.202356s
Starting Epoch 9:
Training loss = 77.906853, training accuracy = 0.973969
Evaluation accuracy = 0.982372, Elapsed Time = 4.191382s
ubuntu@ip-172-31-39-137:~/incubator-singa/examples/autograd$ python3 resnet.py
Start intialization............
100%|███████████████████████████████████████████████████████████████████████| 100/100 [01:26<00:00,  1.14it/s]
Throughput = 36.89267491263885 per second
Total=0.8673808574676514, forward=0.2684857630729675, softmax=0.0027115750312805176, backward=0.5961835193634033, sgd=0.03734057664871216
```

2. This is the old result before improving the momentum SGD:
```
ubuntu@ip-172-31-39-137:~/incubator-singa/examples/autograd$ python3 mnist_cnn.py
Starting Epoch 0:
Training loss = 581.382263, training accuracy = 0.794974
Evaluation accuracy = 0.934495, Elapsed Time = 5.541576s
Starting Epoch 1:
Training loss = 233.281906, training accuracy = 0.920808
Evaluation accuracy = 0.953025, Elapsed Time = 5.492121s
Starting Epoch 2:
Training loss = 169.505447, training accuracy = 0.943503
Evaluation accuracy = 0.971454, Elapsed Time = 5.493372s
Starting Epoch 3:
Training loss = 136.643906, training accuracy = 0.954309
Evaluation accuracy = 0.975761, Elapsed Time = 5.513660s
Starting Epoch 4:
Training loss = 116.743042, training accuracy = 0.960963
Evaluation accuracy = 0.979968, Elapsed Time = 5.526858s
Starting Epoch 5:
Training loss = 103.864464, training accuracy = 0.965732
Evaluation accuracy = 0.979667, Elapsed Time = 5.513694s
Starting Epoch 6:
Training loss = 94.542282, training accuracy = 0.968550
Evaluation accuracy = 0.975461, Elapsed Time = 5.520474s
Starting Epoch 7:
Training loss = 87.548050, training accuracy = 0.971368
Evaluation accuracy = 0.980970, Elapsed Time = 5.535038s
Starting Epoch 8:
Training loss = 83.162071, training accuracy = 0.971485
Evaluation accuracy = 0.975661, Elapsed Time = 5.536836s
Starting Epoch 9:
Training loss = 78.447533, training accuracy = 0.974570
Evaluation accuracy = 0.982772, Elapsed Time = 5.547574s
ubuntu@ip-172-31-39-137:~/incubator-singa/examples/autograd$ python3 resnet.py
Start intialization............
100%|███████████████████████████████████████████████████████████████████████| 100/100 [01:49<00:00,  1.11s/it]
Throughput = 29.05542749993395 per second
Total=1.101343286037445, forward=0.270987823009491, softmax=0.0029543495178222657, backward=0.8274011135101318, sgd=0.3130151700973511
```

From above two sets of results (1) and (2), we can see that the new momentum SGD is much faster after fusing the small operations.

### CODE REVIEW

**Comment by chrishkchris** [comment]
Next, I further improve the time performance by using in-place elementwise multiply in many functions such as ReLU. Here is the result:

```
ubuntu@ip-172-31-39-137:~/incubator-singa/examples/autograd$ python3 mnist_cnn.py
Starting Epoch 0:
Training loss = 585.555542, training accuracy = 0.793390
Evaluation accuracy = 0.939303, Elapsed Time = 4.206943s
Starting Epoch 1:
Training loss = 234.893921, training accuracy = 0.922409
Evaluation accuracy = 0.955729, Elapsed Time = 4.101450s
Starting Epoch 2:
Training loss = 169.515244, training accuracy = 0.943286
Evaluation accuracy = 0.970252, Elapsed Time = 4.104907s
Starting Epoch 3:
Training loss = 136.331894, training accuracy = 0.954442
Evaluation accuracy = 0.968450, Elapsed Time = 4.115959s
Starting Epoch 4:
Training loss = 118.268318, training accuracy = 0.960512
Evaluation accuracy = 0.971755, Elapsed Time = 4.117009s
Starting Epoch 5:
Training loss = 104.006439, training accuracy = 0.965732
Evaluation accuracy = 0.978866, Elapsed Time = 4.117350s
Starting Epoch 6:
Training loss = 93.860809, training accuracy = 0.969067
Evaluation accuracy = 0.977464, Elapsed Time = 4.106471s
Starting Epoch 7:
Training loss = 88.009178, training accuracy = 0.970251
Evaluation accuracy = 0.982873, Elapsed Time = 4.116037s
Starting Epoch 8:
Training loss = 81.978348, training accuracy = 0.972802
Evaluation accuracy = 0.983974, Elapsed Time = 4.121274s
Starting Epoch 9:
Training loss = 75.998878, training accuracy = 0.974103
Evaluation accuracy = 0.982272, Elapsed Time = 4.122591s
```

**Comment by chrishkchris** [comment]
Finally, I test the distributed training in AWS p2.x8large, after adding the Sync() in the SGD loop of resnet.py and resnet_dist.py.
The speed up of using 8 GPUs is now 7.21x, but this is compared without real data feeding. 
See the following throughput comparison in resnet.py and resnet_dist.py:

```
ubuntu@ip-172-31-28-231:~/incubator-singa/examples/autograd$ python3 resnet.py
Start intialization............
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [01:23<00:00,  1.19it/s]
Throughput = 38.13589358185999 per second
Total=0.8391045022010803, forward=0.26401839971542357, softmax=0.0020227289199829103, backward=0.5730633735656739, sgd=0.016838366985321044

ubuntu@ip-172-31-28-231:~/incubator-singa/examples/autograd$ /home/ubuntu/mpich-3.3/build/bin/mpiexec --hostfile host_file python3 resnet_dist.py
Start intialization...........
100%|██████████| 100/100 [01:33<00:00,  1.08it/s]
Throughput = 274.9947180123401 per second
Total=0.9309269714355469, forward=0.2690380573272705, softmax=0.0021610450744628906, backward=0.6597278690338135, sgd=0.10374969005584717
```

File: src/core/tensor/tensor_math_cuda.h [hunk 1]
```diff
@@ -324,12 +324,8 @@ void EltwiseMult<float, lang::Cuda>(const Tensor& in,
                                     const float x, Tensor* out, Context* ctx) {
   const float* inPtr = static_cast<const float*>(in.block()->data());
   float* outPtr = static_cast<float*>(out->block()->mutable_data());
-
-  float alpha = x, beta = 0.0;
-  check_cudnn(cudnnAddTensor(ctx->cudnn_handle,
-                             (void*)(&alpha), generate_tensor_nd_desc(in), inPtr,
-                             (void*)(&beta), generate_tensor_nd_desc(*out), outPtr
-                            ));
+  const size_t num = in.Size();
```
Comment (nudles) [inline src/core/tensor/tensor_math_cuda.h]: is cuda::mult faster than cudnnAddTensor?

Comment (chrishkchris) [reply src/core/tensor/tensor_math_cuda.h]: Yes, must be. It is because cudnnAddTensor add two tensors alphaX+betaY and write the result to Y. Meanwhile, cuda::mult is for purely elementwise multiply. The computational cost of cudnnAddTensor is hence much higher than cuda::mult. We do not need to use a tensor add function to perform elementwise multiply purpose for the consideration of computation cost.

Comment (chrishkchris) [reply src/core/tensor/tensor_math_cuda.h]: In other word, we only need alphaX but not alphaX+betaY (while setting beta=0). The first case has a less computation cost than the latter.
